ledger/narwhal/compact-batch-certificate/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+        // Return the compact batch certificate.
+        Ok(Self { certificate_id, compact_batch_header, signatures })
+    }
+
+    /// Initializes a new compact batch certificate from a batch certificate.
+    pub fn from_batch_certificate(batch_certificate: &BatchCertificate<N>) -> Result<Self> {
+        // Construct the compact batch header.
+        let compact_batch_header = CompactBatchHeader::from_batch_header(batch_certificate.batch_header())?;
+
+        // Construct the signatures.
+        let signatures = batch_certificate
+            .signatures()
+            .zip_eq(batch_certificate.timestamps())
+            .map(|(signature, timestamp)| (*signature, timestamp))
+            .collect::<IndexMap<_, _>>();
+
+        Self::from(batch_certificate.certificate_id(), compact_batch_header, signatures)
+    }
+}
+
+impl<N: Network> CompactBatchCertificate<N> {
+    /// Returns the certificate ID.
+    pub const fn certificate_id(&self) -> Field<N> {
+        self.certificate_id
+    }
+
+    /// Returns the compact batch header.
+    pub const fn compact_batch_header(&self) -> &CompactBatchHeader<N> {
+        &self.compact_batch_header
+    }
+
+    /// Returns the batch ID.
+    pub const fn batch_id(&self) -> Field<N> {
+        self.compact_batch_header.batch_id()
+    }
+
+    /// Returns the author.
+    pub const fn author(&self) -> Address<N> {
+        self.compact_batch_header.author()
+    }
+
+    /// Returns the round.
+    pub const fn round(&self) -> u64 {
+        self.compact_batch_header.round()
+    }
+
+    /// Returns the transmission IDs.
+    pub const fn transmission_ids_map(&self) -> &IndexSet<(TransmissionType, u32)> {
+        self.compact_batch_header.transmission_ids_map()
+    }
+
+    /// Returns the batch certificate IDs for the previous round.
+    pub const fn previous_certificate_ids(&self) -> &IndexSet<Field<N>> {
+        self.compact_batch_header.previous_certificate_ids()
+    }
+
+    /// Returns the median timestamp of the batch ID from the committee.
+    pub fn median_timestamp(&self) -> i64 {
+        let mut timestamps =
+            self.timestamps().chain([self.compact_batch_header.timestamp()].into_iter()).collect::<Vec<_>>();
+        timestamps.sort_unstable();
+        timestamps[timestamps.len() / 2]
+    }
+
+    /// Returns the timestamps of the batch ID from the committee.
+    pub fn timestamps(&self) -> impl '_ + ExactSizeIterator<Item = i64> {
+        self.signatures.values().copied()
+    }
+
+    /// Returns the signatures of the batch ID from the committee.
+    pub fn signatures(&self) -> impl ExactSizeIterator<Item = &Signature<N>> {
+        self.signatures.keys()
+    }
+}
+
+impl<N: Network> Hash for CompactBatchCertificate<N> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.compact_batch_header.batch_id().hash(state);
+        (self.signatures.len() as u64).hash(state);
```

Comment:
nit: it's probably not necessary to hash the number of signatures if all of them get hashed individually as well